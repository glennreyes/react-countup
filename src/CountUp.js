import PropTypes from 'prop-types';
import React, { useEffect, createRef, useRef } from 'react';
import warning from 'warning';
import { createCountUpInstance } from './common';

const arePropsEqual = (prevProps, nextProps) => {
  const hasCertainPropsEqual =
    prevProps.duration === nextProps.duration ||
    prevProps.end === nextProps.end ||
    prevProps.start === nextProps.start;

  return hasCertainPropsEqual || !nextProps.redraw;
};

const CountUp = React.memo(props => {
  const containerRef = createRef();
  const timer = useRef();
  const instance = useRef();

  const createInstance = () => {
    if (typeof props.children === 'function') {
      // Warn when user didn't use containerRef at all
      warning(
        containerRef.current &&
          (containerRef.current instanceof HTMLElement ||
            containerRef.current instanceof SVGTextElement),
        `Couldn't find attached element to hook the CountUp instance into! Try to attach "containerRef" from the render prop to a an HTMLElement, eg. <span ref={containerRef} />.`,
      );
    }
    return createCountUpInstance(containerRef.current, props);
  };

  useEffect(() => {
    const { children, delay } = props;
    instance.current = createInstance();

    // Don't invoke start if component is used as a render prop
    if (typeof children === 'function' && delay !== 0) return;

    // Otherwise just start immediately
    start();

    return () => {
      if (timer.current) {
        clearTimeout(timer.current);
      }
      instance.current.reset();
    };
  }, [props.duration, props.start]);

  useEffect(() => {
    // Only end value has changed, so reset and and re-animate with the updated
    // end value.

    // Don't invoke start if component is used as a render prop
    if (typeof children === 'function' && props.delay !== 0) return;

    instance.current.reset();
    instance.current.update(props.end);
  }, [props.end]);

  const pauseResume = () => {
    const { onPauseResume } = props;

    instance.current && instance.current.pauseResume();

    onPauseResume({ reset, start, update });
  };

  const reset = () => {
    const { onReset } = props;

    onReset({ pauseResume, start, update });
  };

  const restart = () => {
    reset();
    start();
  };

  const update = newEnd => {
    const { onUpdate } = props;

    instance.current && instance.current.update(newEnd);

    onUpdate({ pauseResume, reset, start });
  };

  const start = () => {
    const { delay, onEnd, onStart } = props;
    const run = () =>
      instance.current &&
      instance.current.start(() =>
        onEnd({ pauseResume, reset, start, update }),
      );

    // Delay start if delay prop is properly set
    if (delay > 0) {
      timer.current = setTimeout(run, delay * 1000);
    } else {
      run();
    }

    onStart({ pauseResume, reset, update });
  };

  const { children, className, style } = props;

  if (typeof children === 'function') {
    return children({
      countUpRef: containerRef,
      pauseResume,
      reset,
      start: restart,
      update,
    });
  }

  return <span className={className} ref={containerRef} style={style} />;
}, arePropsEqual);

CountUp.propTypes = {
  decimal: PropTypes.string,
  decimals: PropTypes.number,
  delay: PropTypes.number,
  easingFn: PropTypes.func,
  end: PropTypes.number.isRequired,
  formattingFn: PropTypes.func,
  onEnd: PropTypes.func,
  onStart: PropTypes.func,
  prefix: PropTypes.string,
  redraw: PropTypes.bool,
  separator: PropTypes.string,
  start: PropTypes.number,
  suffix: PropTypes.string,
  style: PropTypes.object,
  useEasing: PropTypes.bool,
};

CountUp.defaultProps = {
  decimal: '.',
  decimals: 0,
  delay: null,
  duration: null,
  easingFn: null,
  formattingFn: null,
  onEnd: () => {},
  onPauseResume: () => {},
  onReset: () => {},
  onStart: () => {},
  onUpdate: () => {},
  prefix: '',
  redraw: false,
  separator: '',
  start: 0,
  suffix: '',
  style: undefined,
  useEasing: true,
};

export default CountUp;
