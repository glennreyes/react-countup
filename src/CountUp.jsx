import React, { PropTypes } from 'react';
import { findDOMNode } from 'react-dom';
import shallowCompare from 'react-addons-shallow-compare';
import Count from 'countup.js';

class CountUp extends React.Component {
  componentDidMount() {
    const {
      start,
      end,
      duration,
      useEasing,
      useGrouping,
      separator,
      decimals,
      decimal,
      prefix,
      suffix,
      callback,
    } = this.props;

    this.state = {
      start,
      end,
      duration,
      useEasing,
      useGrouping,
      separator,
      decimals,
      decimal,
      prefix,
      suffix,
      callback,
    };

    this.startAnimation();
  }

  shouldComponentUpdate(nextProps, nextState) {
    return nextProps.redraw
      ? true
      : shallowCompare(this, nextProps, nextState);
  }

  componentDidUpdate() {
    this.startAnimation();
  }

  startAnimation() {
    const {
      start,
      end,
      duration,
      useEasing,
      useGrouping,
      separator,
      decimals,
      decimal,
      prefix,
      suffix,
      callback,
    } = this.props;

    const countup = new Count(findDOMNode(this), start, end, decimals, duration, {
      useEasing,
      useGrouping,
      separator,
      decimal,
      prefix,
      suffix,
    });

    countup.start(callback);
  }

  render() {
    const { className, start, style } = this.props;

    return <span className={className} style={style}>{start}</span>;
  }
}

CountUp.defaultProps = {
  start: 0,
  end: 2016,
  decimals: 0,
  duration: 4,
  useEasing: true,
  useGrouping: false,
  separator: ',',
  decimal: '.',
  prefix: '',
  suffix: '',
  redraw: false,
};

CountUp.propTypes = {
  className: PropTypes.node,
  style: PropTypes.object,
  start: PropTypes.number.isRequired,
  end: PropTypes.number.isRequired,
  decimals: PropTypes.number,
  duration: PropTypes.number,
  useEasing: PropTypes.bool,
  useGrouping: PropTypes.bool,
  separator: PropTypes.string,
  decimal: PropTypes.string,
  prefix: PropTypes.string,
  suffix: PropTypes.string,
  callback: PropTypes.func,
  redraw: PropTypes.bool,
};

export default CountUp;
