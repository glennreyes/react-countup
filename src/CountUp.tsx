import React, { Component, CSSProperties } from 'react';
import warning from 'warning';
import { createCountUpInstance } from './common';
import { CountUp as CountUpJs } from 'countup.js';
import { CallbackProps, CommonProps, RenderCounterProps } from './types';

export interface CountUpProps extends CommonProps, CallbackProps {
  className?: string;
  redraw?: boolean;
  preserveValue?: boolean;
  children?: (props: RenderCounterProps) => React.ReactNode;
  style?: CSSProperties;
}

class CountUp extends Component<CountUpProps> {
  private instance: CountUpJs | undefined;
  private timeoutId: ReturnType<typeof setTimeout> | undefined;

  static defaultProps = {
    decimal: '.',
    decimals: 0,
    delay: null,
    duration: null,
    easingFn: null,
    formattingFn: null,
    onEnd: () => {},
    onPauseResume: () => {},
    onReset: () => {},
    onStart: () => {},
    onUpdate: () => {},
    prefix: '',
    redraw: false,
    separator: '',
    start: 0,
    startOnMount: true,
    suffix: '',
    style: undefined,
    useEasing: true,
    preserveValue: false,
  };

  componentDidMount() {
    const { children, delay } = this.props;
    this.instance = this.createInstance();

    // Don't invoke start if component is used as a render prop
    if (typeof children === 'function' && delay !== 0) return;

    // Otherwise just start immediately
    this.start();
  }

  checkProps = (updatedProps: CountUpProps) => {
    const {
      start,
      suffix,
      prefix,
      redraw,
      duration,
      separator,
      decimals,
      decimal,
      className,
      formattingFn
    } = this.props;

    const hasPropsChanged =
      duration !== updatedProps.duration ||
      start !== updatedProps.start ||
      suffix !== updatedProps.suffix ||
      prefix !== updatedProps.prefix ||
      separator !== updatedProps.separator ||
      decimals !== updatedProps.decimals ||
      decimal !== updatedProps.decimal ||
      className !== updatedProps.className ||
      formattingFn !== updatedProps.formattingFn;

    return hasPropsChanged || redraw;
  };

  shouldComponentUpdate(nextProps: CountUpProps) {
    const { end } = this.props;
    return this.checkProps(nextProps) || end !== nextProps.end;
  }

  componentDidUpdate(prevProps: CountUpProps) {
    // If duration, suffix, prefix, separator or start has changed
    // there's no way to update the values.
    // So we need to re-create the CountUp instance in order to
    // restart it.
    const { end, preserveValue } = this.props;

    if (this.checkProps(prevProps)) {
      this.instance?.reset();
      this.instance = this.createInstance();
      this.start();
    }

    // Only end value has changed, so reset and and re-animate with the updated
    // end value.
    if (end !== prevProps.end) {
      if (!preserveValue) {
        this.instance?.reset();
      }
      this.instance?.update(end);
    }
  }

  componentWillUnmount() {
    if (this.timeoutId) {
      clearTimeout(this.timeoutId);
    }
    // instance.target is incorrectly marked private by typescript
    if (this.instance && (this.instance as any).target) {
      this.instance.reset();
    }
  }

  createInstance = () => {
    if (typeof this.props.children === 'function') {
      // Warn when user didn't use containerRef at all
      warning(
        this.containerRef.current instanceof Element,
        `Couldn't find attached element to hook the CountUp instance into! Try to attach "containerRef" from the render prop to a an Element, eg. <span ref={containerRef} />.`,
      );
    }
    return createCountUpInstance(this.containerRef.current, this.props);
  };

  pauseResume = () => {
    const { reset, restart: start, update } = this;
    const { onPauseResume } = this.props;

    this.instance?.pauseResume();

    onPauseResume?.({ reset, start, update });
  };

  reset = () => {
    const { pauseResume, restart: start, update } = this;
    const { onReset } = this.props;

    this.instance?.reset();

    onReset?.({ pauseResume, start, update });
  };

  restart = () => {
    this.reset();
    this.start();
  };

  start = () => {
    const { pauseResume, reset, restart: start, update } = this;
    const { delay, onEnd, onStart } = this.props;
    const run = () =>
      this.instance?.start(() => onEnd?.({ pauseResume, reset, start, update }));

    // Delay start if delay prop is properly set
    if (delay && delay > 0) {
      this.timeoutId = setTimeout(run, delay * 1000);
    } else {
      run();
    }

    onStart?.({ pauseResume, reset, update });
  };

  update = (newEnd: string | number) => {
    const { pauseResume, reset, restart: start } = this;
    const { onUpdate } = this.props;

    this.instance?.update(newEnd);

    onUpdate?.({ pauseResume, reset, start });
  };

  containerRef = React.createRef<any>();

  render() {
    const { children, className, style } = this.props;
    const { containerRef, pauseResume, reset, restart, update } = this;

    if (typeof children === 'function') {
      return children({
        countUpRef: containerRef,
        pauseResume,
        reset,
        start: restart,
        update,
      });
    }

    return <span className={className} ref={containerRef} style={style} />;
  }
}

export default CountUp;
