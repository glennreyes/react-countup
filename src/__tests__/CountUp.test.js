import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import CountUp from '../index';

afterEach(cleanup);

describe('CountUp component', () => {
  it('renders start value correctly', () => {
    const { container } = render(<CountUp end={10} />);

    expect(container).toMatchSnapshot();
  });

  it('re-renders change of start value correctly', () => {
    const { container, rerender } = render(<CountUp end={10} />);

    rerender(<CountUp start={5} end={10} />);

    expect(container).toMatchSnapshot();
  });

  it('re-renders change of duration value correctly', () => {
    const { container, rerender } = render(<CountUp duration={1} end={10} />);

    rerender(<CountUp duration={2} end={10} />);

    expect(container).toMatchSnapshot();
  });

  it('clear previous counter when duration changed', (done) => {
    const { container, rerender } = render(<CountUp duration={1} end={100} />);

    rerender(<CountUp duration={0.5} end={10} />);

    setTimeout(() => {
      const span = container.firstChild;
      expect(span.textContent).toEqual('10');
      done();
    }, 1200);
  });

  it('re-renders when suffix changes', (done) => {
    const { container, rerender } = render(<CountUp duration={1} end={1} />);

    rerender(<CountUp duration={1} end={30} suffix=" sec" />);

    setTimeout(() => {
      const span = container.firstChild;
      expect(span.textContent).toEqual('30 sec');
      done();
    }, 1200);
  });

  it('re-renders when the separator changes', (done) => {
    const { container, rerender } = render(
      <CountUp duration={1} end={1} separator="" />,
    );

    rerender(<CountUp duration={1} end={3000} separator=" " />);

    setTimeout(() => {
      const span = container.firstChild;
      expect(span.textContent).toEqual('3 000');
      done();
    }, 1200);
  });

  it('re-renders when the decimals changes', (done) => {
    const { container, rerender } = render(
      <CountUp duration={1} end={1} decimals={0} />,
    );

    rerender(<CountUp duration={1} end={10.5} decimals={1} />);

    setTimeout(() => {
      const span = container.firstChild;
      expect(span.textContent).toEqual('10.5');
      done();
    }, 1200);
  });

  it('re-renders when the decimal changes', (done) => {
    const { container, rerender } = render(<CountUp duration={1} end={1} />);

    rerender(<CountUp duration={1} end={10.5} decimals={1} decimal="." />);

    setTimeout(() => {
      const span = container.firstChild;
      expect(span.textContent).toEqual('10.5');
      done();
    }, 1200);
  });

  it('re-renders when the prefix changes', (done) => {
    const { container, rerender } = render(
      <CountUp duration={1} end={1} prefix="" />,
    );

    rerender(<CountUp duration={1} end={30} prefix="->" />);

    setTimeout(() => {
      const span = container.firstChild;
      expect(span.textContent).toEqual('->30');
      done();
    }, 1200);
  });

  it('re-renders change of end value correctly', () => {
    const { container, rerender } = render(<CountUp end={10} />);

    rerender(<CountUp end={5} />);

    expect(container).toMatchSnapshot();
  });

  it('re-renders with redraw={false} correctly', () => {
    const { container, rerender } = render(<CountUp redraw={false} end={10} />);

    rerender(<CountUp redraw={false} end={10} />);

    expect(container).toMatchSnapshot();
  });

  it('renders with delay correctly', () => {
    const { container } = render(<CountUp delay={1} end={10} />);

    expect(container).toMatchSnapshot();
  });

  it('renders with delay as a render prop component correctly', () => {
    const { container } = render(
      <CountUp delay={1} end={10}>
        {({ countUpRef }) => <div ref={countUpRef} />}
      </CountUp>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders as a render prop component correctly', () => {
    const { container } = render(
      <CountUp end={10}>
        {({ countUpRef }) => <div ref={countUpRef} />}
      </CountUp>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders with autostart correctly', () => {
    const { container } = render(
      <CountUp autostart end={10}>
        {({ countUpRef }) => <div ref={countUpRef} />}
      </CountUp>,
    );

    expect(container).toMatchSnapshot();
  });

  it('re-renders with redraw={true} correctly', () => {
    const onStart = jest.fn();
    const { rerender } = render(<CountUp end={10} onStart={onStart} redraw />);

    rerender(<CountUp end={10} onStart={onStart} redraw />);
    expect(onStart).toHaveBeenCalledTimes(2);
  });

  it('does not reset if preserveValue is true', (done) => {
    const { container, rerender } = render(
      <CountUp duration={1} end={10} preserveValue />,
    );

    setTimeout(() => {
      rerender(<CountUp duration={1} end={20} preserveValue />);
      const span = container.firstChild;
      expect(span.textContent).toEqual('10');
      done();
    }, 1000);
  });

  it('does not reset if preserveValue is true and suffix is set', (done) => {
    const { container, rerender } = render(
      <CountUp duration={1} end={10} suffix="%" preserveValue />,
    );

    setTimeout(() => {
      rerender(<CountUp duration={1} end={20} suffix="%" preserveValue />);
      const span = container.firstChild;
      expect(span.textContent).toEqual('10%');
      done();
    }, 1000);
  });

  it('does not reset if preserveValue is true and prefix is set', (done) => {
    const { container, rerender } = render(
      <CountUp duration={1} end={10} prefix="->" preserveValue />,
    );

    setTimeout(() => {
      rerender(<CountUp duration={1} end={20} prefix="->" preserveValue />);
      const span = container.firstChild;
      expect(span.textContent).toEqual('->10');
      done();
    }, 1000);
  });

  it('calls start correctly', () => {
    const spy = {};

    const { container } = render(
      <CountUp end={10}>
        {({ countUpRef, start }) => {
          spy.start = start;
          jest.spyOn(spy, 'start');
          return <button onClick={spy.start} ref={countUpRef} />;
        }}
      </CountUp>,
    );

    fireEvent.click(container.firstElementChild);

    expect(spy.start).toHaveBeenCalled();
  });

  it('calls pauseResume correctly', () => {
    const spy = {};

    const { container } = render(
      <CountUp end={10}>
        {({ countUpRef, pauseResume }) => {
          spy.pauseResume = pauseResume;
          jest.spyOn(spy, 'pauseResume');
          return <button onClick={spy.pauseResume} ref={countUpRef} />;
        }}
      </CountUp>,
    );

    fireEvent.click(container.firstElementChild);

    expect(spy.pauseResume).toHaveBeenCalled();
  });

  it('calls update correctly', () => {
    const spy = {};

    const { container } = render(
      <CountUp end={10}>
        {({ countUpRef, update }) => {
          spy.update = update;
          jest.spyOn(spy, 'update');
          return <button onClick={spy.update} ref={countUpRef} />;
        }}
      </CountUp>,
    );

    fireEvent.click(container.firstElementChild);

    expect(spy.update).toHaveBeenCalled();
  });

  it('calls reset correctly', () => {
    const spy = {};

    const { container } = render(
      <CountUp end={10}>
        {({ countUpRef, reset }) => {
          spy.reset = reset;
          jest.spyOn(spy, 'reset');
          return <button onClick={spy.reset} ref={countUpRef} />;
        }}
      </CountUp>,
    );

    fireEvent.click(container.firstElementChild);

    expect(spy.reset).toHaveBeenCalled();
  });

  it('calls pauseResume in onStart callback correctly', () => {
    const spy = {};

    render(
      <CountUp
        end={10}
        onStart={({ pauseResume }) => {
          spy.pauseResume = pauseResume;
          jest.spyOn(spy, 'pauseResume');
        }}
      />,
    );

    spy.pauseResume();

    expect(spy.pauseResume).toHaveBeenCalled();
  });

  it('throws warning if ref not attached to a component', () => {
    console.error = jest.fn();
    jest.spyOn(console, 'error');

    render(<CountUp end={10}>{() => <div />}</CountUp>);

    expect(console.error).toBeCalled();
  });
});
