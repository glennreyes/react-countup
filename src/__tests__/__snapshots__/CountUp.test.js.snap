// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountUp component re-renders change of duration value correctly 1`] = `
<div>
  <span>
    0
  </span>
</div>
`;

exports[`CountUp component re-renders change of end value correctly 1`] = `
<div>
  <span>
    0
  </span>
</div>
`;

exports[`CountUp component re-renders change of start value correctly 1`] = `
<div>
  <span>
    10
  </span>
</div>
`;

exports[`CountUp component re-renders with redraw={false} correctly 1`] = `
<div>
  <span>
    10
  </span>
</div>
`;

exports[`CountUp component renders as a render prop component correctly 1`] = `
<div>
  <div>
    0
  </div>
</div>
`;

exports[`CountUp component renders start value correctly 1`] = `
<div>
  <span>
    10
  </span>
</div>
`;

exports[`CountUp component renders with autostart correctly 1`] = `
<div>
  <div>
    0
  </div>
</div>
`;

exports[`CountUp component renders with delay as a render prop component correctly 1`] = `
<div>
  <div>
    0
  </div>
</div>
`;

exports[`CountUp component renders with delay correctly 1`] = `
<div>
  <span>
    0
  </span>
</div>
`;
